doc/api/util.md

```diff
@@ -3305,6 +3305,29 @@ util.isUndefined(null);
 // Returns: false
 ```
 
+### `util.loadEnvFile(path)`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1.1 - Active Development
+
+* `path` {string | undefined}
```

Comment:
Personally I would find it much more useful if this API was `parseEnvFile` and accepted a string, the contents of a `.env` file. This would mean the typical usage would involve two steps: a `readFile` call to get the `.env` file contents, and then passing those contents into `parseEnvFile`. So a little more verbose, but we would gain the ability to make any edits we may want to make in between the two calls; and/or we could generate the `.env` file dynamically, or pull it from a nontraditional source like within a .zip file or via a network call.

Another option is that this is just an `.env` file _parser,_ and doesn’t set anything. So the full equivalent to what you have now would be:

```js
const envFileContents = await readFile('.env', 'utf8')
const parsedEnvironmentVariables = parseEnvFile(envFileContents)
process.env = { ...process.env, ...parsedEnvironmentVariables }
```

Yes this is even more verbose, but it allows maximum flexibility in terms of being able to customize at any point (before parsing, or after parsing but before loading). It also makes clear how `NODE_OPTIONS` behaves; the same as it does now if you define `process.env.NODE_OPTIONS` at runtime.

Reply:
Unfortunately, I disagree. I don't see the value in exposing parse function. My reasoning are:

- I don't see a clear use-case
- Returning object from a C++ function is discouraged

If this is the goal, we could just add a parameter like `{ raw: true }` and return object instead of updating process.env

Reply:
Even if we don’t have it return an object, I think we should separate out the “read file from disk” and “parse string” aspects. As in, allow the input to the API be the string _contents_ of the file, not the path to the file. It’s very annoying to work with APIs that only accept file paths as input when I want that input to be dynamic; I end up needing to write a temp file to disk and then pass the path to that temp file into the API. We know that environment variables are an area where there’s some user demand for dynamism, like for loading “common to all environments” variables plus dev variables, say; it’s easier to deal with that if the merged result doesn’t need to be written to disk first.

Reply:
I have a hard time imagining in which case you would want to pass something that's not a file when you can already change env variables without that API 🤔

Reply:
> I have a hard time imagining in which case you would want to pass something that’s not a file when you can already change env variables without that API 🤔

Loading a `.env` file from something like HashiCorp Vault or AWS Secrets Manager, for example.

```js
const envFileContents = await vaultClient.read('.env')
process.loadEnvFile(envFileContents)
```

Yes I can _set_ environment variables without this API, but I can’t _parse the file_ without either this new API or some library like `dotenv`, or writing my own parser.

Reply:
Would it be acceptable to add an optional parameter for this? I don't want to make the general usage harder by supporting a small percentage of users (referencing HashiCorp vault or AWS Secrets Manager)

```js
process.loadEnvFile(envFileContents, { mode: 'raw' })
```

Reply:
I feel like the API is just simpler overall if it can accept file contents as input. Then we don’t need any options at all; not `raw`, not `force`. And it supports more use cases such as the Secrets Manager or Vault ones, or encrypted `.env` files, etc.

```js
const envFileContents = await readFile(new URL(‘.env’, import.meta.url))
process.loadEnvFile(envFileContents)
```

Reply:
I'm -1 on your suggestion.

For example, if we support array of urls in the future, something like `loadEnvFile(['.development.env', '.env']`, I don't see how it would apply to your proposed solution. But regardless, I think we should provide a really simple API, for a really simple functionality.

Reply:
> For example, if we support array of urls in the future, something like `loadEnvFile(['.development.env', '.env']`, I don’t see how it would apply to your proposed solution.

```js
const envFile = await readFile(new URL('.env', import.meta.url), 'utf8')
const envFileDevelopment = await readFile(new URL('.development.env', import.meta.url), 'utf8')
process.loadEnvFile([envFile, envFileDevelopment].join('\n'))
```